asserts/asserts.go

```diff
@@ -0,0 +1,285 @@
+// where:
+//
+//    HEADER is a set of header lines separated by "\n"
+//    BODY can be arbitrary,
+//    SIGNATURE is composed like:
+//      SIG-TYPE " " BASE64-SIG-PACKET
+//
+//    SIG-TYPE is the type/format expected for the base64
+//    encoded signature value, for now only "openpgp" is supported
+//
+// An header line looks like:
+//
+//   NAME ": " VALUE
+//
+// for sanity NAME is expected to match headerNameSanity regexp.
+//
+// The following headers are mandatory:
+//
+//   type
+//   authority-id (the signer id)
+//   revision (a positive int)
+//   body-length (int expected to be equal to the length of BODY)
+//
+func Decode(serializedAssertion []byte) (Assertion, error) {
+	contentSignatureSplit := bytes.LastIndex(serializedAssertion, nlnl)
+	if contentSignatureSplit == -1 {
+		return nil, fmt.Errorf("assertion content/signature separator not found")
+	}
+	content := serializedAssertion[:contentSignatureSplit]
+	signature := serializedAssertion[contentSignatureSplit+2:]
+
+	headersBodySplit := bytes.Index(content, nlnl)
+	var body, head []byte
+	if headersBodySplit == -1 {
+		head = content
+	} else {
+		body = content[headersBodySplit+2:]
+		if len(body) == 0 {
+			body = nil
+		}
+		head = content[:headersBodySplit]
+	}
+
+	headers, err := parseHeaders(head)
+	if err != nil {
+		return nil, fmt.Errorf("parsing assertion headers: %v", err)
+	}
+
+	if len(signature) == 0 {
+		return nil, fmt.Errorf("empty assertion signature")
+	}
+
+	sigtypeSigpacketSplit := bytes.IndexByte(signature, ' ')
+	if sigtypeSigpacketSplit == -1 {
+		return nil, fmt.Errorf("could not split the assertion signature into type and base64 packet")
+	}
+
+	sigtype := string(signature[:sigtypeSigpacketSplit])
+	sigpacket := signature[sigtypeSigpacketSplit+1:]
+	if len(sigpacket) == 0 {
+		return nil, fmt.Errorf("missing assertion signature packet")
+	}
+	decodedSigpacket := make([]byte, base64.StdEncoding.DecodedLen(len(sigpacket)))
+	n, err := base64.StdEncoding.Decode(decodedSigpacket[:], sigpacket)
+	if err != nil {
+		return nil, fmt.Errorf("could not base64 decode the assertion signature packet")
+	}
+	sigpacket = decodedSigpacket[:n]
+
+	return buildAssertion(headers, body, content, sigtype, sigpacket)
+}
+
+func buildAssertion(headers map[string]string, body, content []byte, sigtype string, sigpacket []byte) (Assertion, error) {
+
+	checkInteger := func(name string) (int, error) {
+		valueStr, ok := headers[name]
+		if !ok {
+			return -1, fmt.Errorf("assertion %v header is mandatory", name)
+		}
+		value, err := strconv.Atoi(valueStr)
+		if err != nil {
+			return -1, fmt.Errorf("assertion %v is not an integer: %v", name, valueStr)
+		}
+		return value, nil
+	}
+
+	length, err := checkInteger("body-length")
+	if err != nil {
+		return nil, err
+	}
+	if length != len(body) {
+		return nil, fmt.Errorf("assertion body length and declared body-length don't match: %v != %v", len(body), length)
+	}
+
+	checkMandatory := func(name string) (string, error) {
```

Comment:
Same as above. This way we'll take 20 lines out of a pretty critical function in the system, making the important logic more readable.